spec/cybersource/base_plugin_spec.rb

```diff
@@ -196,6 +196,83 @@
     end
   end
 
+  def stub_gateway_for_invoice_header(invoice_match_status)
+    ::ActiveMerchant::Billing::CyberSourceGateway.any_instance.stub(:ssl_post) do |host, request_body|
+      case(invoice_match_status)
+      when :none
+        request_body.should_not match('<invoiceHeader>')
+      when :all
+        request_body.should match('<invoiceHeader>\n        <merchantDescriptor>Ray Qiu               </merchantDescriptor>\n        <merchantDescriptorContact>650-888-3161</merchantDescriptorContact>\n      </invoiceHeader>')
+      when :except_authorize
+        if request_body.index('ccAuthService').present?
+          request_body.should_not match('<invoiceHeader>\n        <amexDataTAA1>Ray Qiu</amexDataTAA1>\n        <amexDataTAA2>6508883161</amexDataTAA2>')
+        else
+          request_body.should match('<invoiceHeader>\n        <amexDataTAA1>Ray Qiu</amexDataTAA1>\n        <amexDataTAA2>6508883161</amexDataTAA2>')
+        end
+      end
+      successful_purchase_response
+    end
+  end
+
+  shared_examples 'full payment' do
+    before do
+      send(add_payment_prpperties, txn_properties, card_type)
+      stub_gateway_for_invoice_header(invoice_match_status)
+    end
+
+    it 'should met expectations' do
+      auth_responses = create_transaction(card_type, :authorize, nil, :PROCESSED, txn_properties, expected_successful_params)
+      capture_responses = create_transaction(card_type, :capture, auth_responses, :PROCESSED, txn_properties, expected_successful_params)
+      create_transaction(card_type, :refund, capture_responses, :PROCESSED, txn_properties, expected_successful_params)
+    end
+  end
+
+  shared_examples 'invoice header example' do
+    let(:card_type){ :visa }
+    let(:txn_properties){ [] }
+
+    context 'while no descriptor provided' do
+      let(:invoice_match_status){ :none }
+
+      context 'visa' do
+        it_behaves_like 'full payment'
+      end
+
+      context 'amex' do
+        let(:card_type){ :amex }
+        it_behaves_like 'full payment'
+      end
+    end
+
+    context 'while descriptor provided' do
+      before{ txn_properties << build_property('merchant_descriptor', {"name"=>"Ray Qiu", "contact"=>"6508883161"}.to_json) }
+
+      context 'visa' do
+        let(:invoice_match_status){ :all }
+        it_behaves_like 'full payment'
+      end
+
+      context 'amex' do
+        let(:card_type){ :amex }
+        let(:invoice_match_status){ :except_authorize }
+        it_behaves_like 'full payment'
+      end
+    end
+  end
+
+  context 'Invoice Header' do
+    let(:add_payment_prpperties){ :add_card_property }
+
+    context 'payments with card' do
+      it_behaves_like 'invoice header example'
```

Comment:
the `let` in line 264 created a default setting of `:add_payment_properties` in all the context within line 263. but since there are only 2 context within context line 263, it looks better if put the `let` into context at line 266.
updated this in the next commit.